// react-session-control/dist/UnauthenticatedSessionControl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("react");
const utils_1 = require("./utils");
class UnauthenticatedSessionControl extends react_1.PureComponent {
    constructor(props) {
        super(props);
        const { tokenChangeDebounceTime } = props;
        this.handleStorageChange = this.handleStorageChange.bind(this);
        this.handleStorageKeyChangeDebounced = utils_1.debounce(this.handleStorageKeyChange.bind(this), tokenChangeDebounceTime);
    }
    componentDidMount() {
        window.addEventListener('storage', this.handleStorageChange);
    }
    componentWillUnmount() {
        window.removeEventListener('storage', this.handleStorageChange);
    }
    handleStorageChange(event) {
        const { key } = event;
        const { storageTokenKey } = this.props;
        if (key === storageTokenKey) {
            this.handleStorageKeyChangeDebounced(event);
        }
    }
    handleStorageKeyChange({ newValue }) {
        const { onLogin } = this.props;
        if (newValue != null) {
            onLogin && onLogin();
        }
    }
    render() {
        return (false);
    }
}
exports.default = UnauthenticatedSessionControl;
UnauthenticatedSessionControl.defaultProps = {
    tokenChangeDebounceTime: 500,
};
